sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1032,6 +1248,109 @@ object PushPredicateThroughJoin extends Rule[LogicalPlan] with PredicateHelper {
 }
 
 /**
+ * Pushes down a subquery, in the form of [[Join LeftSemi/LeftAnti]] operator
+ * to the left or right side of a join below.
+ */
+object PushLeftSemiLeftAntiThroughJoin extends Rule[LogicalPlan] with PredicateHelper {
```

Comment:
Pushing down LeftSemi/LeftAnti through a join is better done as part of the cost-based join reordering. This code is put here to preserve the existing behaviour.